src/runtime_symbol_lookup.cc

```diff
@@ -31,86 +38,171 @@ FILE *RuntimeSymbolLookup::perfmaps_open(int pid,
+      symbol_table[existing]._demangle_name = symbol;
+      symbol_table[existing]._symname = symbol;
+    }
+  }
+
+  return true;
 }
 
-void RuntimeSymbolLookup::fill_perfmap_from_file(int pid, SymbolMap &symbol_map,
-                                                 SymbolTable &symbol_table) {
+DDRes RuntimeSymbolLookup::fill_from_jitdump(std::string_view jitdump_path,
+                                             pid_t pid, SymbolMap &symbol_map,
+                                             SymbolTable &symbol_table) {
+  char buf[1024];
+  auto n = snprintf(buf, 1024, "%s/proc/%d/root%s", _path_to_proc.c_str(), pid,
+                    jitdump_path.data());
+  if (n >= 1024) { // unable to snprintf everything
+    DDRES_RETURN_ERROR_LOG(DD_WHAT_JIT, "Unable to create path to jitdump");
+  }
+
+  JITDump jitdump;
+  DDRes res = jitdump_read(std::string_view(buf, n), jitdump);
+  if (IsDDResNotOK(res) && res._what == DD_WHAT_NO_JIT_FILE) {
+    // retry different path
+    res = jitdump_read(jitdump_path, jitdump);
+    if (IsDDResFatal(res)) {
+      // Stop if fatal error
+      return res;
+    }
+  }
+
+  for (const JITRecordCodeLoad &code_load : jitdump.code_load) {
+    insert_or_replace(code_load.func_name, code_load.code_addr,
+                      code_load.code_size, symbol_map, symbol_table);
+  }
+  // todo we can add file and inlined functions with debug info
+  return ddres_init();
+}
+
+bool RuntimeSymbolLookup::should_skip_symbol(std::string_view symbol) const {
+  // we could consider making this more efficient if the table grows
+  for (const auto &el : _ignored_symbols_start) {
+    if (symbol.starts_with(el)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+DDRes RuntimeSymbolLookup::fill_from_perfmap(int pid, SymbolMap &symbol_map,
+                                             SymbolTable &symbol_table) {
   FILE *pmf = perfmaps_open(pid, "/tmp");
-  symbol_map.clear();
   if (pmf == nullptr) {
-    // Add a single fake symbol to avoid bouncing
-    symbol_map.emplace(0, SymbolSpan());
-    LG_DBG("No runtime symbols (PID%d)", pid);
-    return;
+    LG_DBG("Unable to read perfmap file (PID%d)", pid);
+    return ddres_error(DD_WHAT_NO_JIT_FILE);
   }
   defer { fclose(pmf); };
 
   LG_DBG("Loading runtime symbols from (PID%d)", pid);
   char *line = NULL;
   size_t sz_buf = 0;
   char buffer[2048];
-  auto it = symbol_map.end();
   while (-1 != getline(&line, &sz_buf, pmf)) {
     char address_buff[33]; // max size of 16 (as it should be hexa for uint64)
     char size_buff[33];
     // Avoid considering any symbols beyond 300 chars
     if (3 !=
-            sscanf(line, "%16s %8s %300[^\t\n]", address_buff, size_buff,
-                   buffer) ||
-        should_skip_symbol(buffer)) {
+        sscanf(line, "%16s %8s %300[^\t\n]", address_buff, size_buff, buffer)) {
       continue;
     }
+    std::string sym = std::string(buffer);
```

Comment:
Why not pass `buffer` directly to `insert_or_replace` ?

Reply:
better indeed